test/unit/atomic/test-atomic-incdec.cpp

```diff
@@ -0,0 +1,192 @@
+                      std::tuple<unsigned long long int, RAJA::auto_atomic>,
+                      std::tuple<unsigned long long int, RAJA::cuda_atomic>
+#endif
+#if defined(RAJA_ENABLE_HIP)
+                      ,
+                      std::tuple<unsigned int, RAJA::auto_atomic>,
+                      std::tuple<unsigned int, RAJA::hip_atomic>,
+                      std::tuple<unsigned long long int, RAJA::auto_atomic>,
+                      std::tuple<unsigned long long int, RAJA::hip_atomic>
+#endif
+                    >;
+
+// Basic Inc Dec
+
+template <typename T>
+class AtomicBasicIncDecUnitTest : public ::testing::Test
+{};
+
+TYPED_TEST_SUITE_P( AtomicBasicIncDecUnitTest );
+
+TYPED_TEST_P( AtomicBasicIncDecUnitTest, BasicIncDecs )
+{
+  using T = typename std::tuple_element<0, TypeParam>::type;
+  using AtomicPolicy = typename std::tuple_element<1, TypeParam>::type;
+
+  // test "wrapping" increment
+  // See: http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicinc
+
+  T inc_init = (T)0;
+  T* inc_result = &inc_init;
+
+  // oldval < val, increment oldval
+  RAJA::atomicInc<AtomicPolicy>(inc_result, (T)1);
+  ASSERT_EQ( inc_result[0], (T)1 );
+
+  // oldval == val, wrap to 0
+  RAJA::atomicInc<AtomicPolicy>(inc_result, (T)1);
+  ASSERT_EQ( inc_result[0], (T)0 );
+
+  // oldval > val, wrap to 0
+  inc_result[0] = (T)2;
+  RAJA::atomicInc<AtomicPolicy>(inc_result, (T)1);
+  ASSERT_EQ( inc_result[0], (T)0 );
+
+  // test "wrapping" decrement
+  // See: http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicdec
+
+  T dec_init = (T)1;
+  T* dec_result = &dec_init;
+
+  // oldval > 0, decrement oldval
+  RAJA::atomicDec<AtomicPolicy>(dec_result, (T)1);
+  ASSERT_EQ( dec_result[0], (T)0 );
+
+  // oldval == 0, wrap to val
+  RAJA::atomicDec<AtomicPolicy>(dec_result, (T)1);
+  ASSERT_EQ( dec_result[0], (T)1 );
+
+  // oldval > val, wrap to val
+  dec_result[0] = (T)3;
+  RAJA::atomicDec<AtomicPolicy>(dec_result, (T)1);
+  ASSERT_EQ( dec_result[0], (T)1 );
+}
+
+REGISTER_TYPED_TEST_SUITE_P( AtomicBasicIncDecUnitTest,
+                             BasicIncDecs
+                           );
+
+INSTANTIATE_TYPED_TEST_SUITE_P( BasicIncDecUnitTest,
+                                AtomicBasicIncDecUnitTest,
+                                unsigned_types
+                              );
+
+
+// Pure CUDA test.
+#if defined(RAJA_ENABLE_CUDA)
+
+using CUDA_unsigned_types =
+    ::testing::Types<
+                      std::tuple<unsigned int, RAJA::auto_atomic>,
+                      std::tuple<unsigned int, RAJA::cuda_atomic>,
+                      std::tuple<unsigned long long int, RAJA::auto_atomic>,
+                      std::tuple<unsigned long long int, RAJA::cuda_atomic>
+                    >;
+
+
+template <typename T>
+class AtomicCUDAIncDecUnitTest : public ::testing::Test
+{};
+
+TYPED_TEST_SUITE_P( AtomicCUDAIncDecUnitTest );
+
+GPU_TYPED_TEST_P( AtomicCUDAIncDecUnitTest, CUDAIncDecs )
```

Comment:
Since these tests are CUDA-specific, does it make sense to switch from a gpu testing pattern to a normal one?

Reply:
The GPU test macros just make a function with your code in it and call it in the test instead of having your code directly in the test. The testing pattern found in test/unit/resource are what I'm referring to as the normal pattern. The normal pattern uses a .cpp.in file and uses cmake to generate a version per backend to avoid code duplication. In the normal testing pattern cartesian product is used in the cpp file to instantiate test templates with combinations of different types.

Reply:
Ahh, that makes sense. I'll go ahead and try to make that change.